lsp-mode.el

```diff
@@ -3326,15 +3338,19 @@ WORKSPACE is the active workspace."
    (lsp--parser-body p) nil
    (lsp--parser-reading-body p) nil))
 
-(defun lsp--read-json (str use-native-json)
-  (let* ((use-native-json (and use-native-json (fboundp 'json-parse-string)))
-         (json-array-type (if use-native-json 'vector 'list))
+(defun lsp--read-json (str)
+  "Read json string STR."
+  (let* ((use-native-json (and lsp-use-native-json (fboundp 'json-parse-string)))
+         (json-array-type (if lsp-json-use-lists 'list 'vector))
          (json-object-type 'hash-table)
          (json-false nil))
     (if use-native-json
         (with-no-warnings
-          (json-parse-string str :object-type 'hash-table
-                             :null-object nil :false-object nil))
+          (with-temp-buffer
+            (json-parse-string str
+                               :object-type 'hash-table
+                               :null-object nil
+                               :false-object :json-false)))
```

Comment:
Switching from `nil` to `:json-false` leads to conditionals with json values to suddenly succeed. For example, `(gethash "resolveProvider" ...)` in `lsp--sort-completions` now returns a truth-value (`:json-false`) for ccls, leading to errors during completion. There may be other similar instances.

Reply:
Thanks for spotting this. We had to evaluate whether to keep the current(incorrect) behaviour of fix all places which do if over a lsp result.

Reply:
I'd vote for `:false`/`:null`, as in the defaults of the emacs27 json library, or even `nil` instead of `:false`, that way you preserve truth value.

Reply:
We already use `:json-false` for false when doing serialization.

Reply:
But that doesn't affect us much as we don't reuse the objects we create during serialization. On the other hand, using a non-`nil` value while working with values we get from the language server would inadvertently result in a lot of broken code. Every `if` and conditional construct would have to be rewritten to check for the new "false", which would just result in inorganic and unidiomatic lisp code, prone to breaking.

Reply:
@vibhavp there are few places we reuse the object, e. g. - lenses, diagnostics, code actions.

> which would just result in inorganic and unidiomatic lisp code, prone to breaking.

I agree that it will be ugly, error-prone, difficult to migrate and support. And I agree with you that it is better to keep it as it is at least for now...

A link to a code that deals with that issue - I am converting the nil to `:json-false` to make it work. 
https://github.com/emacs-lsp/lsp-java/blob/master/lsp-java-boot.el#L68